eav/queryset.py

```diff
@@ -1,47 +1,273 @@
-#    it under the terms of the GNU Lesser General Public License as published
-#    by the Free Software Foundation, either version 3 of the License, or
-#    (at your option) any later version.
-#
-#    This software is distributed in the hope that it will be useful,
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU Lesser General Public License for more details.
-#
-#    You should have received a copy of the GNU Lesser General Public License
-#    along with EAV-Django.  If not, see <http://gnu.org/licenses/>.
+'''
+Queryset module.
 
+This module contains custom EavQuerySet class used for overriding
+relational operators and pure functions for rewriting Q-expressions.
+Q-expressions need to be rewritten for two reasons:
+
+    1. In order to hide implementation from the user and provide easy to use
+       syntax sugar, i.e.::
+
+           Supplier.objects.filter(eav__city__startswith='New')
+
+       instead of::
+
+           city_values = Value.objects.filter(value__text__startswith='New')
+           Supplier.objects.filter(eav_values__in=city_values)
+
+       For details see: ``eav_filter``.
+
+    2. To ensure that Q-expression tree is compiled to valid SQL.
+       For details see: ``rewrite_q_expr``.
+'''
+
+import sys
+from functools import reduce, wraps
+
+from django.db import models
+from django.db.models import Q
 from django.db.models.query import QuerySet
 
-from .decorators import eav_filter
+from .models import Attribute, Value
+
+
+def is_eav_and_leaf(expr, gr_name):
+    '''
+    Checks whether Q-expression is an EAV AND leaf.
+    '''
+    return (getattr(expr, 'connector', None) == 'AND' and
+            len(expr.children) == 1 and
+            expr.children[0][0] in ['pk__in', '{}__in'.format(gr_name)])
+
+
+def rewrite_q_expr(model_cls, expr):
+    '''
+    Rewrites Q-expression to safe form, in order to ensure that
+    generated SQL is valid.
+
+    Suppose we have the following Q-expression:
+
+    └── OR
+         ├── AND
+         │    └── eav_values__in [1, 2, 3]
+         └── AND (1)
+              ├── AND
+              │    └── eav_values__in [4, 5]
+              └── AND
+                   └── eav_values__in [6, 7, 8]
+
+    All EAV values are stored in a single table. Therefore, INNER JOIN
+    generated for the AND-expression (1) will always fail, i.e.
+    single row in a eav_values table cannot be both in two disjoint sets at
+    the same time (and the whole point of using AND, usually, is two have
+    two different sets). Therefore, we must paritially rewrite the
+    expression so that the generated SQL is valid::
+
+    └── OR
+         ├── AND
+         │    └── eav_values__in [1, 2, 3]
+         └── AND
+              └── pk__in [1, 2]
+
+    This is done by merging dangerous AND's and substituting them with
+    explicit ``pk__in`` filter, where pks are taken from evaluted
+    Q-expr branch.
```

Comment:
From my experience I know that having IN(?...) statements with hundreds or thousands variables inside can slow down compilation considerably or even fail parsing. Could it become a problematic case here (especially with OR operators)?

Let's assume an ecommerce store. As far as I understand the description above, the expressions are being evaluated as a intermediary step. Is it possible that one could implicitly and unknowingly write a seemingly innocent expression that during the evaluation returns 99% of store's products and pass it to the next filtering step?